### PR TITLE
Implemented shape inference for Gather

### PR DESCRIPTION
**Pull request by ryotatomioka**
For `data_shape = [d0, d1, ..., d{r-1}]` and`indices_shape = [s0, s1, ..., s{q-1}]`, the output shape is `[d0, ..., d{axis-1}, s0, ..., s{q-1}, d{axis+1}, ..., d{r-1}]`.

### CODE REVIEW

File: onnx/defs/tensor/defs.cc [hunk 1]
```diff
@@ -685,9 +685,14 @@ ONNX_OPERATOR_SET_SCHEMA(
             return;
           }
 
-          int r = ctx.getInputType(0)->tensor_type().shape().dim_size();
-          int q = ctx.getInputType(1)->tensor_type().shape().dim_size();
-
+          auto& data_shape = ctx.getInputType(0)->tensor_type().shape();
```
Comment (snnn) [inline onnx/defs/tensor/defs.cc]: Don't use auto unless it increases readability

File: onnx/defs/tensor/defs.cc [hunk 2]
```diff
@@ -685,9 +685,14 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+          auto& indices_shape = ctx.getInputType(1)->tensor_type().shape();
```
Comment (snnn) [inline onnx/defs/tensor/defs.cc]: Is it possible it doesn't have a shape?

File: onnx/defs/tensor/defs.cc [hunk 3]
```diff
@@ -697,7 +702,11 @@ ONNX_OPERATOR_SET_SCHEMA(
             ctx.getOutputType(0)
                 ->mutable_tensor_type()
                 ->mutable_shape()
-                ->add_dim();
+                ->add_dim()
+                ->set_dim_value(
+                (i < axis) ? data_shape.dim(i).dim_value() : // nb: axis < r
```
Comment (snnn) [inline onnx/defs/tensor/defs.cc]: Is it possible it doesn't have a dim_value?

File: onnx/defs/tensor/defs.cc [hunk 4]
```diff
@@ -685,9 +685,14 @@ ONNX_OPERATOR_SET_SCHEMA(
...
           }
 
-          int r = ctx.getInputType(0)->tensor_type().shape().dim_size();
-          int q = ctx.getInputType(1)->tensor_type().shape().dim_size();
-
+          auto& data_shape = ctx.getInputType(0)->tensor_type().shape();
+          auto& indices_shape = ctx.getInputType(1)->tensor_type().shape();
+          int r = data_shape.dim_size();
+          int q = indices_shape.dim_size();
+          int axis = static_cast<int>(getAttribute(ctx, "axis", 0));
```
Comment (snnn) [inline onnx/defs/tensor/defs.cc]: Is this cast safe? why?

File: onnx/defs/tensor/defs.cc [hunk 5]
```diff
@@ -685,9 +685,14 @@ ONNX_OPERATOR_SET_SCHEMA(
             return;
           }
 
-          int r = ctx.getInputType(0)->tensor_type().shape().dim_size();
-          int q = ctx.getInputType(1)->tensor_type().shape().dim_size();
-
+          auto& data_shape = ctx.getInputType(0)->tensor_type().shape();
+          auto& indices_shape = ctx.getInputType(1)->tensor_type().shape();
+          int r = data_shape.dim_size();
```
Comment (snnn) [inline onnx/defs/tensor/defs.cc]: check if r >= 1?

File: onnx/defs/tensor/defs.cc (hunk 4 above)
Comment (snnn) [reply onnx/defs/tensor/defs.cc]: check if axis is in [-r, r-1]?

File: onnx/defs/tensor/defs.cc [hunk 6]
```diff
@@ -685,9 +685,14 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+          int q = indices_shape.dim_size();
+          int axis = static_cast<int>(getAttribute(ctx, "axis", 0));
+          if (axis < 0) {
+            axis = r + axis;
```
Comment (snnn) [inline onnx/defs/tensor/defs.cc]: If you like, you can use axis+=r;

File: onnx/defs/tensor/defs.cc (hunk 1 above)
Comment (ryotatomioka) [reply onnx/defs/tensor/defs.cc]: Fixed.

File: onnx/defs/tensor/defs.cc (hunk 5 above)
Comment (ryotatomioka) [reply onnx/defs/tensor/defs.cc]: Done.

File: onnx/defs/tensor/defs.cc (hunk 2 above)
Comment (ryotatomioka) [reply onnx/defs/tensor/defs.cc]: I believe this is already checked by `hasNInputShapes(ctx, 2)`

File: onnx/defs/tensor/defs.cc (hunk 4 above)
Comment (ryotatomioka) [reply onnx/defs/tensor/defs.cc]: Fixed.

File: onnx/defs/tensor/defs.cc (hunk 6 above)
Comment (ryotatomioka) [reply onnx/defs/tensor/defs.cc]: Done.

File: onnx/defs/tensor/defs.cc (hunk 3 above)
Comment (ryotatomioka) [reply onnx/defs/tensor/defs.cc]: I don't think so. See my comments in 2ff5e38342eca39fc1334d8e242fac0cf681431a

File: onnx/defs/tensor/defs.cc (hunk 4 above)
Comment (ryotatomioka) [reply onnx/defs/tensor/defs.cc]: The cast is safe because it is from `int64_t` to `int`

File: onnx/defs/tensor/defs.cc [hunk 7]
```diff
@@ -697,7 +707,11 @@ ONNX_OPERATOR_SET_SCHEMA(
             ctx.getOutputType(0)
                 ->mutable_tensor_type()
                 ->mutable_shape()
-                ->add_dim();
+                ->add_dim()
+                ->set_dim_value(
+                (i < axis) ? data_shape.dim(i).dim_value() :                             // i < axis < r
```
Comment (snnn) [inline onnx/defs/tensor/defs.cc]: Is dim_value always available?

Comment (ryotatomioka) [reply onnx/defs/tensor/defs.cc]: Since `i < axis < r` I think it exists, or did I misunderstand?

Comment (snnn) [reply onnx/defs/tensor/defs.cc]: If you open up the protobuf defs. this field is optional.

**Comment by ryotatomioka** [comment]
Hi @snnn could you have a final look at the PR?

**Comment by snnn** [comment]
@gramalingam Could you please take a look? Thanks.

File: onnx/defs/tensor/defs.cc [hunk 8]
```diff
@@ -684,20 +684,43 @@ ONNX_OPERATOR_SET_SCHEMA(
...
-            ctx.getOutputType(0)
+            auto* newdim = ctx.getOutputType(0)
                 ->mutable_tensor_type()
                 ->mutable_shape()
                 ->add_dim();
+            const auto& dim =
+                (i < axis) ? data_shape.dim(i) :                             // i < axis < r
+                (i >= axis && i < axis + q) ? indices_shape.dim(i - axis) :  // i - axis < q
+                data_shape.dim(i - q + 1);                                   // i < out_rank < q + r - 1
+            if (dim.has_dim_value()) {
```
Comment (gramalingam) [inline onnx/defs/tensor/defs.cc]: Wouldn't "*newdim = dim" have the same effect as these lines of code?

**Comment by ryotatomioka** [comment]
Thank you @gramalingam. This PR is ready to be merged.

**Comment by ryotatomioka** [comment]
Could @gramalingam or @houseroad check and merge this PR? It is becoming painful to keeping it up-to-date with the master...

**Comment by gramalingam** [comment]
Close and open to retrigger CI

**Comment by gramalingam** [comment]
The appveyor CI failure looks strange … and I am unable to trigger the CI to happen again.

**Comment by ryotatomioka** [comment]
@gramalingam I've merged master and the CI seems happy.

**Pull request merged**